File: backends/Contracts/Handler/IHandler.cs
using D2.Contracts.Result;

namespace D2.Contracts.Handler;

/// <summary>
/// Represents a handler that process an input of type <see cref="TInput"/> and produces an output
/// of type <see cref="TOutput"/>.
/// </summary>
///
/// <typeparam name="TInput">
/// The type of input the handler processes.
/// </typeparam>
/// <typeparam name="TOutput">
/// The type of output the handler produces.
/// </typeparam>
public interface IHandler<in TInput, TOutput>
{
    /// <summary>
    /// Handles the given input asynchronously and produces an output.
    /// </summary>
    ///
    /// <param name="input">
    /// The input to be processed by the handler.
    /// </param>
    /// <param name="ct">
    /// A cancellation token to cancel the operation.
    /// </param>
    /// <param name="options">
    /// Optional handler options to customize the handling behavior.
    /// </param>
    ///
    /// <returns>
    /// A <see cref="ValueTask{D2Result}"/> representing the asynchronous operation,
    /// containing the result of the handling process.
    /// </returns>
    public ValueTask<D2Result<TOutput?>> HandleAsync(
        TInput input,
        CancellationToken ct = default,
        HandlerOptions? options = null
    );
}

File: backends/node/shared/handler/src/i-handler.ts
import type { D2Result } from "@d2/result";
import type { HandlerOptions } from "./handler-options.js";
import type { RedactionSpec } from "./redaction-spec.js";

/**
 * Handler interface that processes an input and produces an output wrapped in D2Result.
 * Mirrors D2.Shared.Handler.IHandler<TInput, TOutput> in .NET.
 */
export interface IHandler<TInput, TOutput> {
  handleAsync(input: TInput, options?: HandlerOptions): Promise<D2Result<TOutput | undefined>>;
  /** Redaction posture for this handler's I/O logging. */
  readonly redaction?: RedactionSpec;
}
